wandb/sdk/internal/sender.py

```diff
@@ -327,6 +327,11 @@ def send_request_defer(self, data):
         elif state == defer.FLUSH_FS:
             if self._fs:
                 # TODO(jhr): now is a good time to output pending output lines
+                if self._pusher:
+                    # pusher generates some events for filestream, so we need
+                    # to join on pusher to make sure that we have all events
+                    # flushed for filestream to finish with
+                    self._pusher.join()
```

Comment:
I think this isnt quite right.

self._pusher.join will block the sender thread.  The join was being handled by the client in the finish() path polling for file updates,  so I think this will break polling when there are lots of files to update.

Reply:
i think the proper fix will require restructuring the state machine a bit to make sure that the file pusher was drained before continuing the state machine to shutting down filestream.

Reply:
I am writing a quick test to verify this.

Reply:
Yep, this test demonstrates the problem:
```
#!/usr/bin/env python
import os
import wandb

run = wandb.init()
fname = run.dir + "/junk"
print("Writing data...", fname)
for num in range(100):
    os.system("dd if=/dev/urandom bs=1024 count=2048 of=%s-%d.dat" % (fname, num))
print("Done writing")
```

The polling gets blocked while pusher is blocked.